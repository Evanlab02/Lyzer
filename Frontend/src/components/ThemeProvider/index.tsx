import { useEffect, useState } from "react";
import { ThemeContext } from "../../hooks/useTheme";

export const ThemeProvider: React.FC<{ children: React.ReactNode }> = ({ children }) => {
	const defaultDark = window.matchMedia("(prefers-color-scheme: dark)").matches;

	const [isDarkMode, setDarkMode] = useState(
		localStorage.getItem("isDarkMode") !== null
			? localStorage.getItem("isDarkMode") === "true"
			: defaultDark
	);

	const toggleTheme = () => {
		setDarkMode((prev) => !prev);
	};

	useEffect(() => {
		const theme = isDarkMode ? "dark" : "light";
		localStorage.setItem("isDarkMode", theme === "dark" ? "true" : "false");
		document.documentElement.setAttribute("data-theme", theme);
	}, [isDarkMode]);

	return (
		<ThemeContext.Provider value={{ isDarkMode, toggleTheme }}>
			{children}
		</ThemeContext.Provider>
	);
};