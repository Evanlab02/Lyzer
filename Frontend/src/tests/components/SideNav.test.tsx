import { vi, expect, it } from "vitest";
import { MemoryRouter } from "react-router-dom";
import { fireEvent, render } from "@testing-library/react";
import ThemeProvider from "../../components/ThemeProvider";
import SideMenu from "../../components/SideNav";


it("SideNav renders closed correctly", async () => {
	const { findByTestId } = render(
		<MemoryRouter initialEntries={["/"]}>
			<ThemeProvider>
				<SideMenu
					onCloseClick={vi.fn()}
					onSideNavBlur={vi.fn()}
					sideNavOpen={false}
					testId="lyzer-side-nav"
				/>
			</ThemeProvider>
		</MemoryRouter>
	);

	const sideNav = await findByTestId("lyzer-side-nav");
	expect(sideNav).toMatchSnapshot();
});

it("SideNav renders open correctly", async () => {
	const { findByTestId } = render(
		<MemoryRouter initialEntries={["/"]}>
			<ThemeProvider>
				<SideMenu
					onCloseClick={vi.fn()}
					onSideNavBlur={vi.fn()}
					sideNavOpen={true}
					testId="lyzer-side-nav"
				/>
			</ThemeProvider>
		</MemoryRouter>
	);

	const sideNav = await findByTestId("lyzer-side-nav");
	expect(sideNav).toMatchSnapshot();
});

it("SideNav calls onCloseClick when menuToggle is clicked", async () => {
	const onCloseClick = vi.fn();
	const { findByTestId } = render(
		<MemoryRouter initialEntries={["/"]}>
			<ThemeProvider>
				<SideMenu
					onCloseClick={onCloseClick}
					onSideNavBlur={vi.fn()}
					sideNavOpen={true}
					testId="lyzer-side-nav"
					menuToggleTestId="lyzer-side-nav-menu-toggle"
				/>
			</ThemeProvider>
		</MemoryRouter>
	);

	const menuToggle = await findByTestId("lyzer-side-nav-menu-toggle");
	fireEvent.click(menuToggle);

	expect(onCloseClick).toHaveBeenCalledTimes(1);
});
