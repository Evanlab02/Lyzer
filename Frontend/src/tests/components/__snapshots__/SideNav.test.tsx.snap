// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SideNav renders closed correctly 1`] = `
<div
  class="side-nav closed"
  data-testid="lyzer-side-nav"
  id="side-menu"
>
  <svg
    class="lucide lucide-arrow-left-to-line closeIcon"
    fill="none"
    height="28"
    stroke="black"
    stroke-linecap="round"
    stroke-linejoin="round"
    stroke-width="2"
    viewBox="0 0 24 24"
    width="28"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M3 19V5"
    />
    <path
      d="m13 6-6 6 6 6"
    />
    <path
      d="M7 12h14"
    />
  </svg>
  <div
    class="links"
    id="links"
  >
    <a
      class="link-container selected"
      data-discover="true"
      href="/"
    >
      <svg
        class="lucide lucide-house"
        fill="none"
        height="24"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M15 21v-8a1 1 0 0 0-1-1h-4a1 1 0 0 0-1 1v8"
        />
        <path
          d="M3 10a2 2 0 0 1 .709-1.528l7-5.999a2 2 0 0 1 2.582 0l7 5.999A2 2 0 0 1 21 10v9a2 2 0 0 1-2 2H5a2 2 0 0 1-2-2z"
        />
      </svg>
      <div>
        OVERVIEW
      </div>
    </a>
    <a
      class="link-container "
      data-discover="true"
      href="/drivers"
    >
      <svg
        class="lucide lucide-user"
        fill="none"
        height="24"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M19 21v-2a4 4 0 0 0-4-4H9a4 4 0 0 0-4 4v2"
        />
        <circle
          cx="12"
          cy="7"
          r="4"
        />
      </svg>
      <div>
        DRIVERS
      </div>
    </a>
  </div>
</div>
`;

exports[`SideNav renders open correctly 1`] = `
<div
  class="side-nav open"
  data-testid="lyzer-side-nav"
  id="side-menu"
>
  <svg
    class="lucide lucide-arrow-left-to-line closeIcon"
    fill="none"
    height="28"
    stroke="black"
    stroke-linecap="round"
    stroke-linejoin="round"
    stroke-width="2"
    viewBox="0 0 24 24"
    width="28"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M3 19V5"
    />
    <path
      d="m13 6-6 6 6 6"
    />
    <path
      d="M7 12h14"
    />
  </svg>
  <div
    class="links"
    id="links"
  >
    <a
      class="link-container selected"
      data-discover="true"
      href="/"
    >
      <svg
        class="lucide lucide-house"
        fill="none"
        height="24"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M15 21v-8a1 1 0 0 0-1-1h-4a1 1 0 0 0-1 1v8"
        />
        <path
          d="M3 10a2 2 0 0 1 .709-1.528l7-5.999a2 2 0 0 1 2.582 0l7 5.999A2 2 0 0 1 21 10v9a2 2 0 0 1-2 2H5a2 2 0 0 1-2-2z"
        />
      </svg>
      <div>
        OVERVIEW
      </div>
    </a>
    <a
      class="link-container "
      data-discover="true"
      href="/drivers"
    >
      <svg
        class="lucide lucide-user"
        fill="none"
        height="24"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M19 21v-2a4 4 0 0 0-4-4H9a4 4 0 0 0-4 4v2"
        />
        <circle
          cx="12"
          cy="7"
          r="4"
        />
      </svg>
      <div>
        DRIVERS
      </div>
    </a>
  </div>
</div>
`;
